Issue/353 improvements for screen readers (Part 2)

zachberry opened the pull request:
Updates screen reader branch to resolve issues listed here https://github.com/ucfopen/Obojobo/pull/292#pullrequestreview-165798768


* Questions are no longer labelled as simply "Question" so now either the button is read ('Try Question' or 'Start Question N of M') or the question is read (if the question has been 'flipped over')
* Exiting post-assessment score modal refocuses on content
* MathEquation now has a screen-readable only element that says "Equation ${LABEL}:" before the equation
* Discussed IRL, but it seems that screen readers (at least Chromevox and VO) read formatted text separately and there doesn't seem to be a way to prevent that, so that has not been fixed.
* Flipping questions now _should_ read the question
* Slight refactor of rendering elements to nav.js
* Added 'contentType' property to nav items which are used when creating aria-labels for nav items
* Nav items (built by navStore) now have a parent property
* Nav items now say which content type they are as well as if that item is currently selected. subLink items say what item the are a part of.
* Updated ModalStore to use Common.page.focus method (instead of element.focus)

zachberry commented:
Discussed IRL - There doesn't seem to be a way that I've found to have a screen reader read the contents of a fieldset when not focused in that fieldset. We don't want to remove the fieldset as having it gains us a lot for accessibility. I posted a question here just in case - https://stackoverflow.com/questions/53052524/is-it-possible-to-force-a-screen-reader-to-read-the-inner-contents-of-a-fieldset

As a compromise I'm going to add some hidden text to explain what is inside the fieldset so that it might clue a user as to what to expect. I'm adding "Multiple choice form with X choices", So when a question is read it will say (for example):

> What is 2+2? Multiple choice form with 2 choices. Pick the correct answer.

zachberry commented:
Updated overview of the contents of this pull request:

An overview of what this PR contains:

* Skip Navigation button added
* Updates stylelint-declaration-use-variable to allow for 'inherit' color values
* Improves keyboard control by making Nav tabbable, Next/Prev buttons tabbable, prevent being able to tab inside questions
* All components now have tabIndex = -1 by default, making them focusable
* focus:component works even if the component is on another page, which is navigated to via actions
* focus:component can now be called to focus on the element in the DOM but not with the visual fadeout
* Navigation now using nav tag instead of div
* Fixed bug where hidden navigation was still being read by a screen reader (Chromevox)
* Focus store no longer uses timeouts to make focus fade-in/fade-out effect work (Using CSS animations instead)
* Focus store view states now constants. 'enter' and 'leave' states removed.
* Closing dialog returns focus to last focused element
* Navigating to a new page focuses you to the content
* viewer:focusOnContent event added
* getDomId method added to OboModel to standardize creation of ids for dom elements
* Fixed bug where Header (Shown when nav closed) would sometimes show "?" instead of the title of the current nav target
* Show Question now focuses on question in both Assessment and Practice
* Added Common.page.focus method which resolves some issues with VoiceOver not reading newly focused element
* Question feedback moved to be next to the selected question
* Can close all dialogs with ESC
* MCAssessment answer choices prefixed with hidden text to explain the number of answer choices inside the form
* Questions are no longer labelled as simply "Question" so now either the button is read ('Try Question' or 'Start Question N of M') or the question is read (if the question has been 'flipped over')
* Exiting post-assessment score modal refocuses on content
* MathEquation now has a screen-readable only element that says "Equation ${LABEL}:" before the equation
* Flipping questions now read the question
* Slight refactor of rendering elements to nav.js
* Added 'contentType' property to nav items which are used when creating aria-labels for nav items
* Nav items (built by navStore) now have a parent property
* Nav items now say which content type they are as well as if that item is currently selected. subLink items say what item the are a part of.
* Updated ModalStore to use Common.page.focus method (instead of element.focus)

zachberry commented:
Additional items to do, discussed IRL:

* [ ] - Esc still seems to not be working
* [ ] - Clicking on a nav item should refocus on the nav, not on the content
* [ ] - focusOnContent should focus on the first item in the content area, not the whole content area
* [ ] - Remove aria-label for KaTeX elements

gitnix commented:
**Suggestion**
Once a question is revealed there is no Try Question button, meaning there is nothing delimiting the start of a question. Maybe the phrase `”Practice Question”` or `”Question”` should be read allowed to denote the start of a question. This would help when navigating back through content that has revealed questions. (It is good that `”Multiple choice form”` or such is read after the question, but I feel having that alone is too generic and overall the questions should have a strictly defined start in the content. 

**Unexpected behavior**
In Example Document go to the first practice question. After the question is revealed and read aloud
`”Try Question button”` is read aloud at the end.

Resuming an attempt dialogue reads as `We el el résumé where you left off`. The reader reads `resume` correctly in other contexts, so just noting in case something could be done to fix this.

For an assessment question - content is read twice upon initial focus. For example,
in the Example Document, I get `"Apple is a type of - multiple choice question - pick the correct answer - region main"` followed immediately by`"Apple is a type of - multiple choice question - pick the correct answer - fruit vegetable"

zachberry commented:
@gitnix Addressing your suggestions:

* Switched to using `&apos;` fixed the 'resume' issue in the resume attempt dialog.
* Questions are using a question aria-label, causing Chromevox at least to read 'Question region'.
* Removed an aria-live="aggressive" property on the Question which I think fixes the problem with questions being read more than once

zachberry commented:
Changes for this latest commit -

* Moves FocusUtil and FocusStore from the Common namespace to the Vie… …
…wer namespace. The functions of the FocusStore now are more targeted towards the viewer and it made more sense to have them in this namespace.

* As a result of the change above OboComponent was moved into the Viewer namespace as well.
* Focusing on a question now focuses on the first item in that question rather than the whole question
* Removed using the raw LaTeX code as an aria-label for MathEquation. It was there to support VoiceOver since VoiceOver doesn't read <math> elements but this downgrades other screenreaders who can read math. As a result we decided not to support VoiceOver in this case since usage is low.
* focusOnContent methods are now available in viewer components. These static methods are called by ViewerApp and should focus the DOM on the start of any content the component contains.
* Fixed a bug causing Questions to be read strangely (caused by aria-live="assertive")
* Removed focusing code from the Module viewer-component. This was not the proper place for this code and it's been moved to ViewerApp. Now any future Modules developed don't have to handle basic DOM focusing.
* Assessment Section viewer component has getCurrentStep as a static method
* `viewer:focusOnContent` event removed in favor of `FocusUtil.focusOnNavTargetContent` (Which will focus on the content of the currently navigated to page/section). FocusStore / FocusUtil is now the primary place to manage what _should_ be focused and has been expanded to also be able to target the navigation (More detail in the comments added in FocusStore).
* The code to remove visual focus has been moved into ViewerApp so that FocusStore no longer is managing dom events. This change means I don't need to attach and reattach events to document as this is now handled in react (with the onFocus event on the ref="container" div in ViewerApp)
* Assessment section child pages changed from functional to stateful components as these now need to listen for an event on mount from the Assessment section (used when focusOnContent is called)
* Navigating to another nav target now focuses on the first item in the page (using focusOnContent) rather than focusing on the whole page. This change (to our understanding) is the more expected behavior
* Clicking on a nav item will focus on the whole Navigation rather than focusing on the content (which again to our understanding is the expected behavior)
* Added NavUtil.isNavOpen

zachberry commented:
@samuel-belcastro The phrase "Correct!" (and "Incorrect") can be customized which was why it says that way, but I agree it's confusing. Tried some alternatives - are these better?

`"$CORRECT_LABEL - You received a SCORE% on this question."`
or
`"$INCORRECT_LABEL - You received a SCORE% on this question."`

So, some examples:

`"Correct! - You received a 100% on this question."`
`"Incorrect - You received a 0% on this question."`

Or with some other labels:

`"You got it! - You received a 100% on this question."`
`"Great job - You received a 100% on this question."`
`"Sorry, nope - You received a 0% on this question."`

zachberry commented on packages/obojobo-document-engine/ObojoboDraft/Chunks/MCAssessment/viewer-component.scss:
```diff
@@ -377,83 +355,57 @@
-				}
-			}
-		}
+	// prettier-ignore
```
Needed, otherwise prettier puts rules on different lines and then stylelint gets upset

zachberry commented:
@gitnix Removed all the commented out code you found and updated those imports to be consistent.
@qwertynerd97 Explanation now should focus on the first item. Focusing on a YouTube element when opening a question now at least says "YouTube Video Region". I'd like to do a lot more once our YouTube chunk uses the YouTube API rather than a simple iframe (for example, we could read out the title of the video).

samuel-belcastro commented:
@zachberry Not sure if you addressed this, but 

```
"$CORRECT_LABEL - You received a SCORE% on this question."
or
"$INCORRECT_LABEL - You received a SCORE% on this question."
```

seem good to me. The other options are also good, but I like that they provide a score.

zachberry commented:
@qwertynerd97 I have a fix for the LTI status issue. For the YouTube issue I think we have to wait for the YT API as there's no way for me to talk to that iframe as far as I know, so deferring to #568. For Math it keeps coming down to this issue where ChromeVox doesn't like to read math unless you explicitly focus on the math. Example, if you use the "try me" ActionButton to focus on the math equation on the first page you can press CTRL+CMD+Plus to then read the equation. I wouldn't expect anyone to know to do that of course. I don't know why ChromeVox does this. We could go back to using aria-labels with the raw LaTeX code to resolve this. That works, but obviously then you short-circuit the screen-reader's math parser and the LaTeX code could be hard to understand.

zachberry commented:
Making a note here - one possible idea to investigate - generating english from a latex expression (see https://link.springer.com/content/pdf/10.1007%2F978-3-540-85980-2_36.pdf )

I went on a small tangent exploring MathJax as an alternative and I still ran into issues. It seems like Chromevox has trouble with reading math. In [this MathJax example](http://jsbin.com/?html=%3C!DOCTYPE%20html%3E%0A%3Chtml%3E%0A%3Chead%3E%0A%20%20%3Cmeta%20charset%3D%22utf-8%22%3E%0A%20%20%3Cmeta%20name%3D%22viewport%22%20content%3D%22width%3Ddevice-width%22%3E%0A%20%20%3Ctitle%3EMathJax%20example%3C%2Ftitle%3E%0A%20%20%3Cscript%20type%3D%22text%2Fjavascript%22%20async%0A%20%20src%3D%22https%3A%2F%2Fcdnjs.cloudflare.com%2Fajax%2Flibs%2Fmathjax%2F2.7.5%2Flatest.js%3Fconfig%3DTeX-MML-AM_CHTML%22%20async%3E%0A%3C%2Fscript%3E%0A%3C%2Fhead%3E%0A%3Cbody%3E%0A%3Cp%3E%0A%20%20When%20%5C(a%20%5Cne%200%5C)%2C%20there%20are%20two%20solutions%20to%20%5C(ax%5E2%20%2B%20bx%20%2B%20c%20%3D%200%5C)%20and%20they%20are%0A%20%20%24%24x%20%3D%20%7B-b%20%5Cpm%20%5Csqrt%7Bb%5E2-4ac%7D%20%5Cover%202a%7D.%24%24%0A%3C%2Fp%3E%0A%3C%2Fbody%3E%0A%3C%2Fhtml%3E&live) it fails to read the "b squared" term in the quadratic formula equation correctly and instead says "b two".

Interested in seeing how JAWS handles these issues.

zachberry commented:
Found this, which attempts to convert MathML to an english description - https://github.com/ryanhemphill/mathml-model

qwertynerd97 commented:
MathSpeak looks like a good solution for screenreaders.  It's possible that this could be paired with the themes the @zachberry developed during hack day to allow users to toggle between MathSpeak and MathML's native description, in order to best support both visually impaired users and users with learning disabilities

zachberry commented:
A few more links via @qwertynerd97 

* http://www.daisy.org/blog/mathspeak-mathml-cloud-and-accessible-math
daisy.org
* https://www.washington.edu/doit/what-mathspeak

zachberry commented:
More projects implementing Mathspeak:

* https://www.npmjs.com/package/speech-rule-engine
* http://benetech.github.io/mathjax/
* https://github.com/benetech/mathjax

gitnix commented:
After answering a practice question, and selecting `Check your answer`, `Try Question again button` is immediately read out. I would expect the result of my submission be read out immediately after selecting `Check your answer` (correct/incorrect). If you navigate back up through the responses you do get `Your correct response blah` when over the selected response, but this requires knowing to backtrack through the responses, and doesn't give immediate feedback.

To get inside the content of modals, you must navigate down to the bottom of all the page content. For example, in `other content types`, `click me to see a message`, you must click down 4 times going through other content to get to the content of the modal.

When taking an assessment, you first encounter `question region main`. You cannot click/select this. You must first navigate down which brings you to the second question `try question` button and then go back up to get to the selectable `try question` button of the first question.

`Click to explain attempt score in detail` doesn't seem explain anything.

gitnix commented:
The first 2 issues listed above (practice questions/modals) occur when using the ChromeVox click. If I use a regular mouse click or the Enter button, things work as expected.

gitnix commented:
The latest commit resolved issue with assessment question focus.
